### PR TITLE
THRIFT-4621 Add THeader for Python

### PR DESCRIPTION
**Pull request by spladug**
Client: py

This implements the same subset of [THeaderProtocol](https://github.com/apache/thrift/blob/master/doc/specs/HeaderFormat.md) that the C++ implementation does. Specifically, it supports TBinaryProtocol and TCompactProtocol payloads and allows those same protocols to come in without headers framed/unframed for backwards compatibility. It also supports the gzip transform.

Both Python 2 and Python 3 pass the full test suite (feature and cross) with C++ thrown in the mix:

```
$ make cross
/usr/bin/python3 test/test.py --retry-count 5 --features .* --skip-known-failures --server cpp,c_glib,,,,,py,py3,,,,,,,,,,,
...
No unexpected failures.
0 failed of 40 tests in total.
/usr/bin/python3 test/test.py --retry-count 5 --skip-known-failures --server cpp,c_glib,,,,,py,py3,,,,,,,,,,, --client cpp,c_glib,,,,,py,py3,,,,,,,,,,, --regex ".*"
...
No unexpected failures.
0 failed of 712 tests in total.
```

The only newly added known failures for this are those using the HTTP transport which I expect to not work with this implementation of THeader.

```diff
--- a/test/known_failures_Linux.json
+++ b/test/known_failures_Linux.json
@@ -83,6 +83,8 @@
   "cpp-py3_compact-accelc_http-ip-ssl",
   "cpp-py3_compact_http-ip",
   "cpp-py3_compact_http-ip-ssl",
+  "cpp-py3_header_http-ip",
+  "cpp-py3_header_http-ip-ssl",
   "cpp-py3_json_http-ip",
   "cpp-py3_json_http-ip-ssl",
   "cpp-py3_multi-accel_http-ip",
@@ -101,6 +103,8 @@
   "cpp-py3_multic-multiac_http-ip-ssl",
   "cpp-py3_multic_http-ip",
   "cpp-py3_multic_http-ip-ssl",
+  "cpp-py3_multih-header_http-ip",
+  "cpp-py3_multih-header_http-ip-ssl",
   "cpp-py3_multij-json_http-ip",
   "cpp-py3_multij-json_http-ip-ssl",
   "cpp-py3_multij_http-ip",
@@ -113,6 +117,8 @@
   "cpp-py_compact-accelc_http-ip-ssl",
   "cpp-py_compact_http-ip",
   "cpp-py_compact_http-ip-ssl",
+  "cpp-py_header_http-ip",
+  "cpp-py_header_http-ip-ssl",
   "cpp-py_json_http-ip",
   "cpp-py_json_http-ip-ssl",
   "cpp-py_multi-accel_http-ip",
@@ -131,6 +137,8 @@
   "cpp-py_multic-multiac_http-ip-ssl",
   "cpp-py_multic_http-ip",
   "cpp-py_multic_http-ip-ssl",
+  "cpp-py_multih-header_http-ip",
+  "cpp-py_multih-header_http-ip-ssl",
   "cpp-py_multij-json_http-ip",
   "cpp-py_multij-json_http-ip-ssl",
   "cpp-py_multij_http-ip",
@@ -375,6 +383,8 @@
   "py-cpp_binary_http-ip-ssl",
   "py-cpp_compact_http-ip",
   "py-cpp_compact_http-ip-ssl",
+  "py-cpp_header_http-ip",
+  "py-cpp_header_http-ip-ssl",
   "py-cpp_json_http-ip",
   "py-cpp_json_http-ip-ssl",
   "py-d_accel-binary_http-ip",
@@ -420,6 +430,8 @@
   "py3-cpp_binary_http-ip-ssl",
   "py3-cpp_compact_http-ip",
   "py3-cpp_compact_http-ip-ssl",
+  "py3-cpp_header_http-ip",
+  "py3-cpp_header_http-ip-ssl",
   "py3-cpp_json_http-ip",
   "py3-cpp_json_http-ip-ssl",
   "py3-d_accel-binary_http-ip",
```

### CODE REVIEW

File: lib/py/src/protocol/THeaderProtocol.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+PROTOCOLS_BY_ID = {
+    THeaderSubprotocolID.BINARY: TBinaryProtocolAccelerated,
+    THeaderSubprotocolID.COMPACT: TCompactProtocolAccelerated,
+}
+
+
+class THeaderProtocol(TProtocolBase):
+    def __init__(self, transport, allowed_client_types=None):
+        if not isinstance(transport, THeaderTransport):
+            transport = THeaderTransport(transport, allowed_client_types)
```
Comment (pacejackson) [inline lib/py/src/protocol/THeaderProtocol.py]: Is this common in the other protocols?

File: lib/py/src/server/TServer.py [hunk 2]
```diff
@@ -76,10 +77,16 @@ def serve(self):
...
+
             itrans = self.inputTransportFactory.getTransport(client)
             otrans = self.outputTransportFactory.getTransport(client)
             iprot = self.inputProtocolFactory.getProtocol(itrans)
-            oprot = self.outputProtocolFactory.getProtocol(otrans)
+
+            if isinstance(self.inputProtocolFactory, THeaderProtocolFactory):
+                oprot = iprot
+            else:
+                oprot = self.outputProtocolFactory.getProtocol(otrans)
```
Comment (pacejackson) [inline lib/py/src/server/TServer.py]: Can you add a comment about why we need these special case for THeader?

**Comment by spladug** [comment]
Here's docs on the header format for info: https://github.com/apache/thrift/blob/master/doc/specs/HeaderFormat.md

File: lib/py/src/transport/THeaderTransport.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            raise TTransportException(
+                TTransportException.SIZE_LIMIT,
+                "Header size is larger than whole frame.",
+            )
+
+        self._protocol_id = readVarint(buffer_transport)
+
+        transforms = []
+        transform_count = readVarint(buffer_transport)
+        for _ in range(transform_count):
```
Comment (pacejackson) [inline lib/py/src/transport/THeaderTransport.py:252]: 🔕 Does this use `six` or have a `_compat` module you can use to use `xrange` in py2?

**Review by pacejackson — APPROVED** [review]
I think this looks good!  I had a couple small comments/suggestions.  Note that I'm not all that familiar with how thrift protocols write/read so while the work done in `THeaderTransport` looks correct to me, it's possible that I missed something (although the test results sound like it's all good)

File: lib/py/src/protocol/THeaderProtocol.py (hunk 1 above)
Comment (spladug) [reply lib/py/src/protocol/THeaderProtocol.py]: Nope, it's a weirdness of how much of the logic has to live in the transport level. THeader violates the usual pluggability quite a bit.

Comment (pacejackson) [reply lib/py/src/protocol/THeaderProtocol.py]: Might be another good place to add a comment.

File: lib/py/src/transport/THeaderTransport.py (hunk 3 above)
Comment (spladug) [reply lib/py/src/transport/THeaderTransport.py:252]: yeah, I'm not sure it's worth it though. Other parts of lib/py just use `range` always and in this particular case we currently only support one transform type anyway so at most it'd be returning a single-element list on py2 which isn't really a big deal.

File: lib/py/src/protocol/THeaderProtocol.py (hunk 1 above)
Comment (spladug) [reply lib/py/src/protocol/THeaderProtocol.py]: fair enough :)

done

File: lib/py/src/transport/THeaderTransport.py (hunk 3 above)
Comment (pacejackson) [reply lib/py/src/transport/THeaderTransport.py:252]: Makes sense, esp if the rest of the library is like that.

**Review by pacejackson — APPROVED** [review]

**Comment by spladug** [comment]
This was merged upstream via https://github.com/apache/thrift/pull/1583

**Pull request closed without merging**